tools/first-address-from-root/src/main.rs

```diff
@@ -0,0 +1,163 @@
+use anyhow::{anyhow, bail, Context};
+use cardano_multiplatform_lib::{
+    address::{BaseAddress, StakeCredential},
+    crypto::Bip32PublicKey,
+};
+use reqwest::{blocking::Client, header::CONTENT_TYPE};
+use std::str::FromStr;
+use structopt::StructOpt;
+
+const STAKING_KEY_INDEX: u32 = 0;
+const EXTERNAL: u32 = 0;
+const CHIMERIC_ACCOUNT_DERIVATION: u32 = 2;
+
+#[derive(Debug)]
+#[repr(u8)]
+enum NetworkId {
+    Testnet = 0,
+    Mainnet = 1,
+}
+
+impl FromStr for NetworkId {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "mainnet" => Ok(NetworkId::Mainnet),
+            "testnet" => Ok(NetworkId::Testnet),
+            _ => bail!("Invalid network id. Should be either mainnet or testnet."),
+        }
+    }
+}
+
+impl std::fmt::Display for NetworkId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let as_str = match self {
+            NetworkId::Testnet => "testnet",
+            NetworkId::Mainnet => "mainnet",
+        };
+
+        write!(f, "{as_str}")
+    }
+}
+
+#[derive(Debug, StructOpt)]
+struct Opt {
+    #[structopt(short, long)]
+    public_key: String,
+
+    #[structopt(short, long)]
+    network: NetworkId,
+}
+
+fn main() -> Result<(), anyhow::Error> {
+    let opt = Opt::from_args();
+
+    let carp_base_url = format!("https://gate.flint-wallet.com/{}/carp", opt.network);
+
+    let pk = hex::decode(opt.public_key)
+        .context("public key should be a valid hex string")
+        .and_then(|bytes| {
+            Bip32PublicKey::from_bytes(&bytes).map_err(|_| anyhow!("invalid public key"))
+        })?;
+
+    let staking_key = pk
+        .derive(CHIMERIC_ACCOUNT_DERIVATION)
+        .and_then(|pk| pk.derive(STAKING_KEY_INDEX))
+        .map_err(|e| anyhow!("couldn't derive staking key. Reason {e}"))?;
+
+    let spending = pk
+        .derive(EXTERNAL)
+        .map_err(|e| anyhow!("couldn't derive external tree: {e}"))?
+        .derive(0)
+        .map_err(|e| anyhow!("couldn't derive first address: {e}"))?;
+
+    let base_address = BaseAddress::new(
+        opt.network as u8,
+        &StakeCredential::from_keyhash(&spending.to_raw_key().hash()),
+        &StakeCredential::from_keyhash(&staking_key.to_raw_key().hash()),
+    );
+
+    // let enterprise = EnterpriseAddress::new(
```

Comment:
do we need to keep that?

Reply:
well... I actually don't know. But probably not.